### PR TITLE
Prevent overriding file info of another file when reimport creates extra files

### LINKED ISSUE #75525: Glb import error
### Godot version

v4.0.stable.official [92bee43ad]

### System information

Windows 10, 64bit

### Issue description

First, please run the reproduction project.
When I run the test scene, It report error like follow and app exited. 
image
And I don't know where was error.

### Steps to reproduce

1. Open the reproduction project.
2. Open scene 'user:\\xd\XActors\Weapon\Gun\Rifle\XRifle.tscn'
3. Assign the scene's Meshinstance's mesh with 'user:\\xd\XActors\Weapon\Gun\Rifle\XRifle01.res'
4. Run the scene
Opps, the scene exited.
GIF

### Minimal reproduction project

[GlbTest.zip](https://github.com/godotengine/godot/files/11120181/GlbTest.zip)

NOTE: The project that I provided was picked up from my owner project. Sorry I can't provide a minimal reproduction project because I don't know which step cause this issue. I was test like this:
Create a project, import the glb file, create meshinstance and assign mesh, but there was not error!

### LINKED ISSUE #77978: Binary mesh resource disappears from MeshInstance3D upon project restart
### Godot version

v4.1.beta.custom_build [72b59325c]

### System information

Windows 10, Vulkan

### Issue description

This has been plaguing me for a while in my main project. I can only partially reproduce it in a MRP:
When placing a binary mesh resource imported from a .glb file into a MeshInstance3D, it will be gone upon restart. It doesn't seem to happen to every resource, but I've had MIs that were fine and then start doing it out of nowhere.
I thought it was going to be related to materials due to things like https://github.com/godotengine/godot/issues/77007 but I can reliably trigger it with the steps listed below.
I have done limited testing on .tres files because I usually use .res for size/performance, but in the MRP .tres doesn't seem to trigger it, so perhaps it doesn't affect those.

Note on the Steps to reproduce: you have to delete the test_cube.res and redo the steps to trigger the bug again, but in my main project the affected resources always disappear.

### Steps to reproduce

1. Double click test_cube.glb
2. Click the mesh (test_cube_Cube_097), set save to file path as: res://test_cube.res
3. Make sure to click the little folder icon and then Save (this seems to make it consistent to proc)
4. Press Reimport
5. Drag the test_cube.res to the Mesh slot of the MeshInstance3D
6. Ctrl+s to save
7. Project -> Reload current project
8. Mesh is no longer in the MeshInstance3D

### Minimal reproduction project

[binary res bug.zip](https://github.com/godotengine/godot/files/11681009/binary.res.bug.zip)

### LINKED ISSUE #80132: MeshInstance3D loses reference to mesh resource
### Godot version

4.1

### System information

Godot v4.1.1.stable.mono - Windows 10.0.19045 - Vulkan (Forward+) - dedicated NVIDIA GeForce RTX 2070 SUPER (NVIDIA; 31.0.15.2802) - AMD Ryzen 9 3900X 12-Core Processor (24 Threads)

### Issue description

Saving a MeshInstance3D with a mesh resource imported from a blender file will lose the reference to the mesh resource when reopening de project.

When testing I have found that the uid referencing the mesh saved in the .tscn is the same as the one int the .blend.import and not an unique one for the specific mesh
uid in .tscn: image
uid in .blend.import: image
I don't know if that is the cause of the error but I have found a way to force it to generate a unique uid:

1. select the mesh in the MeshInstance3D and "Make Unique"
2. reassign the mesh resource to the MeshInstance3D

now the .tscn has a unique uid for the mesh and remains referenced when reopening the project (this happens randomly sometimes too when working on the scene)
image

Related: [ #57053](https://github.com/godotengine/godot/issues/57053)

### Steps to reproduce

1. import a .blend file with a mesh (if the .blend has multiple mesh the error doesn't seems reproduce)
2. open the blender importer, set mesh paths and reimport
3. create a new MeshInstance3D node and assign the created .res as a mesh resource
4. save the scene and restart Godot.

### Minimal reproduction project

[mesh_test.zip](https://github.com/godotengine/godot/files/12232647/mesh_test.zip)

### LINKED ISSUE #83154: glTF import won't update extracted texture
### Godot version

4.1.2 stable

### System information

Godot v4.1.2.stable - Windows 10.0.19045 - Vulkan (Forward+) - dedicated Quadro P620 () - 11th Gen Intel(R) Core(TM) i9-11900K @ 3.50GHz (16 Threads)

### Issue description

The textures packed into glb exports won't always update when the inherited glb is updated from blender.  It might use an old texture, appear textureless, or sometimes just work and I simply cannot figure out what i am doing wrong.

### Steps to reproduce

- open project in godot and browse empty testcube folder
- copy testcube-1.glb from zip root into godot_project\testcube as testcube.glb
- switch back to godot and import glb as new inherited scene and save tscn in testcube folder. glb may appear with red "x", but reimporting fixes it?
- copy testcube-2.glb from zip root into godot_project\testcube as testcube.glb
- switch back to godot, image texture should change with some errors in output log
- copy testcube-1.glb from zip root into godot_project\testcube as testcube.glb
- cube might appear with changed texture or appear white/textureless. it has been difficult to get this to predictably happen

white-cube-errors

errors in output log appear: 

```
  Can't find file 'uid://w87lq3421wmi::::res://testcube/testcube_godette.png'.
  Attempt to disconnect a nonexistent connection from 'testcube:<Node3D#2317655010421>'. Signal: 'replacing_by', callable: 'EditorNode::set_edited_scene'.
  Can't find file 'uid://c5oarbyq0oygi::::res://testcube/testcube_godot.png'.
  Attempt to disconnect a nonexistent connection from 'testcube:<Node3D#2486517690152>'. Signal: 'replacing_by', callable: 'EditorNode::set_edited_scene'.
```

blend file notes: Blender 3.6.0, this is the default cube and default material with an image texture of a png file, the only difference between the two glb exports is which files is used. glbs are exported "limited to selected objects" with only the cube selected.

### Minimal reproduction project

[glb-import-issue.zip](https://github.com/godotengine/godot/files/12872041/glb-import-issue.zip)

### PR DESCRIPTION
**Pull request by Listwon**
Fixes #80132, fixes #75525, fixes #77978, fixes #83154

Special thanks to @MetRiko and @gonzelvis for helping me figure out the exact repro steps. They had issues with `*.glb` files and repro turned out to be exactly the same as in #80132.

`EditorFileSystem::_reimport_file` calls `_find_file(p_file, &fs, cpos)` to get the file index in `filesystem` (`fs`). The problem is - if reimport creates extra fiiles, the `filesystem` gets invalidated and the next usage of `cpos` is incorrect - `_reimport_file` updates wrong file info leading to duplicated uid, dependencies etc. The uid inside created file is still correct at this point, but the data in `fs->files[cpos]` is invalid, so saving the scene after assigning the affected resource causes the resource to save with wrong uid.

Adding `_find_file(p_file, &fs, cpos)` just before the next  line  with `cpos` refreshes the `cpos` index so now it points to a proper file in `fs->files`.

### CODE REVIEW

**Review by KoBeWi — APPROVED** [review]
The issue is confusing to reproduce and the fix is not obvious. I'd add a comment next to the new `_find_file()` explaining why is it called.

Other than that looks ok.

**Comment by MetRiko** [comment]
@KoBeWi 
Here is some my thoughts about this PR to show it might be little more important than it looks like.

Every time I was trying to do something with Godot 4.x (personal projects, game jams etc.), I was always reaching the point where random scenes were getting corrupted. It was a constant thing for me especially in the 3D projects.

I was trying to fix scenes with the "Fix dependency" there never was any error.
I also looked into these "corrupted" .tscn with the text editor but then again.. I wasn't able to spot anything wrong. 

After long investigation of this issue (with the huge help of @gonzalveis) we were able to found the exact source of those problems and 100% working repro steps that could lead into corrupted scenes (and more). 

TL;DR 
This PR fixes problem which indirectly could cause:
- disappearing set references to resources inside nodes,
- invisible meshes on the scene (but the instance of such .tscn has visible mesh),
- scene is invalid/corrupt but works in the runtime,
- duplicated UIDs for two different resources,
- resource UID is changed on the saving project,
- restarting projects breaks resource references,
- and many more issues related to importing/creating resources (especially the extracted resources from .glb files).

I don't have time to examine it further, but after what I learnt when I was investigating this bug I think this probably fixes the following issues (you can see most of those issues are related to glTF and/or MeshInstance3D):
#80132, #68672, #77978, #75970, #63513, #83575, #79545, #75525, #83200, #83154, #82275, #80536, #79779, #81671, #77978, #79275, #85850, #73916, #54774

Those above would be worth to check first but there might be the chance this also can fix some of those issues:
#84981, #82652, #82422, #82702, #76236, #77007, #76691

And for testing purposes, the easiest repro steps for the bug, that this PR is fixing, are the following:
0. Prepare any .glb. If you don't have any then exported default cube from Blender will work for the repro too:
- Open Blender, select default cube, export it as .glb (with enabled option "Limit to: Selected Objects" in the export settings).
1. Create empty project with Godot 4.2, create scene with any type of node and save it.
2. Put .glb inside the project -> open it with double click to open glTF importing window
3. Export meshes from this .glb in any way you want, preferable is: 
- Press "Actions..." -> "Set Mesh Save Paths" -> "Select Current Folder" (.tres/.res setting doesn't matter) -> "Set Paths" -> "Reimport"
4. Double click on the any created meshes (exported from .glb) in the filesystem to refresh editor (DON'T SAVE ANYTHING!)
- Now you can see one of those meshes has the SCENE icon instead of MESH icon (what is weird)
5. Check UID of this mesh with wrong icon ("RMB-> Copy UID" and paste it somewhere)
6. Add this mesh to the scene (for example by drag and dropping it on the scene)
7. Save the scene with 3D node (just press ctrl+s)
- You can see icon for this mesh resource has been changed to the valid mesh icon
8. Check UID of this mesh resource again and compare with previous one: 
- **UID for this resource has been changed!**
- **This UID is exactly the same as the .glb file has!**

Now using such corrupted mesh with duplicated UID anywhere will lead to many different problems like corrupting scenes.

**Comment by KoBeWi** [comment]
@MetRiko I checked these issues and was able to confirm 2 more that this PR fixes (already added them in the OP). I also closed 2 issues as already resolved/duplicate.
Other issues seem unrelated, though in some cases I think this PR does not fix things automatically; it's more to prevent the bugs from happening, which makes it more difficult to test as a fix.

**Comment by akien-mga** [comment]
Thanks! Great work tracking this down :1st_place_medal:

**Comment by MetRiko** [comment]
@KoBeWi 
> Other issues seem unrelated, though in some cases I think this PR does not fix things automatically

That's exactly what I meant by:
> this PR fixes problem which indirectly could cause

and by 
> Now using such corrupted mesh with duplicated UID anywhere will lead to many different problems like corrupting scenes.

This will not fix directly problems caused by duplicated UIDs but will prevent UID's duplication. There is no reason in "fixing" bugs caused by using resources with duplicated UIDs because engine is not even suppose to have duplicated UIDs anywhere at all. Each resource should always have one unique UID right? So there is no point in fixing those anomalies, however it would be nice to have a good readable error/message that will point out duplicated UIDs and how to fix it.

**Comment by lyuma** [comment]
Wow!! Thanks for fixing this! I think you and I may have discovered the problem independently [around the same time, Wednesday last week](https://twitter.com/Lyuma2d/status/1732284046519214225). If I had submitted my discovery as a pull request, it might have saved you the time of doing this. Based on my diagnosis/understanding, this fix is correct.

The issue only affected files beginning with the letters "r" through "z" due to another problem in EditorFileSystem where it tries to keep directories sorted, but compares all strings against the full path starting with "res://", so most files were simply appended to the end of the directory, which would then not affect the `cpos`. (My workaround was to extract all embedded resources into a subdirectory, so as to not affect `cpos`.)

> because engine is not even suppose to have duplicated UIDs anywhere at all

What happens if the user manually copies and pastes a file elsewhere in the project, and this file has a particular UID embedded in its .res header?

**Comment by MetRiko** [comment]
@lyuma 
> What happens if the user manually copies and pastes a file elsewhere in the project, and this file has a particular UID embedded in its .res header?

Let's say you have file Entity.tscn and you will create EntityCopy.tscn by duplicating it outside Godot editor (in OS file explorer or with commands). In that case Entity.tscn and EntityCopy.tscn will have the same UID.

Now if you will:
- remove UID part from EntityCopy.tscn with text editor,
- open EntityCopy.tscn inside Godot and press ctrl+s,
then EntityCopy.tscn will get the same UID as before (still duplicated).

However if you will:
 - remove UID part from EntityCopy.tscn with text editor,
 - press "Project->Reload Current Project",
 - open EntityCopy.tscn inside Godot and press ctrl+s,
then new UID for EntityCopy.tscn will be generated.

For the references.. let's say you used EntityCopy.tscn inside World.tscn when EntityCopy has still duplicated UID.
If you will open World.tscn and save it (ctrl+s), Godot will try to update `ext_resource` header automatically (you don't have to reload project). Depend on the situation, the header will be updated differently. There are 5 cases worth to mention:
- `uid=*, path=*` - nothing will be updated.
- `---, path=*` - uid will be taken from the file pointed by path.
- `uid="", path=*` - uid will be taken from the file pointed by path.
- `uid=*, ---` - SCENE CORRUPTED! You cannot open such scene within the Godot.
- `uid=*, path=""` - path will be determined based on the uid.

(where `uid=*` means property with proper path, `uid=""`, property with empty path and `---`, property removed entirely).

I think this should be raised as a separated issue/proposal.

**Comment by YuriSizov** [comment]
Cherry-picked for 4.2.2.

**Comment by YuriSizov** [comment]
Cherry-picked for 4.1.4.
